crates/api_crud/src/site/create.rs

```diff
@@ -157,3 +141,197 @@ impl PerformCrud for CreateSite {
+        rate_limit_register_per_second: None,
+        rate_limit_image: None,
+        rate_limit_image_per_second: None,
+        rate_limit_comment: None,
+        rate_limit_comment_per_second: None,
+        rate_limit_search: None,
+        rate_limit_search_per_second: None,
+        federation_enabled: None,
+        federation_debug: None,
+        federation_worker_count: None,
+        captcha_enabled: None,
+        captcha_difficulty: None,
+        allowed_instances: None,
+        blocked_instances: None,
+        taglines: None,
+        registration_mode: None,
+        auth: Default::default(),
+      }
+    }
+
+    let invalid_payloads = [
+      (
+        false,
+        &Some(String::from("(foo|bar)")),
+        &create_payload(
+          String::from("foo site_name"),
+          None::<String>,
+          None::<String>,
+          None::<String>,
+        ),
+        "slurs",
+      ),
+      (
+        false,
+        &Some(String::from("(foo|bar)")),
+        &create_payload(
+          String::from("zeta site_name"),
+          None::<String>,
+          None::<String>,
+          Some(String::from("(zeta|alpha)")),
+        ),
+        "slurs",
+      ),
+      (
+        true,
+        &None::<String>,
+        &create_payload(
+          String::from("site_name"),
+          None::<String>,
+          None::<String>,
+          None::<String>,
+        ),
+        "site_already_exists",
+      ),
+    ];
+    let valid_payloads = [
+      (
+        false,
+        &None::<String>,
+        &create_payload(
+          String::from("site_name"),
+          None::<String>,
+          None::<String>,
+          None::<String>,
+        ),
+      ),
+      (
+        false,
+        &None::<String>,
+        &create_payload(
+          String::from("site_name"),
+          Some(String::new()),
+          Some(String::new()),
+          None::<String>,
+        ),
+      ),
+      (
+        false,
+        &Some(String::from("(foo|bar)")),
+        &create_payload(
+          String::from("foo site_name"),
+          Some(String::new()),
+          Some(String::new()),
+          Some(String::new()),
+        ),
+      ),
+    ];
+
+    invalid_payloads.iter().enumerate().for_each(
+      |(idx, &(is_site_setup, local_slur_filter_regex, create_site, expected_err))| {
+        match validate_create_payload(is_site_setup, local_slur_filter_regex.clone(), create_site) {
+          Ok(_) => {
+            panic!(
+            "Got Ok, but validation should have failed with error: {} for invalid_payloads.nth({})",
+            expected_err, idx
+          )
+          }
+          Err(error) => {
+            assert!(
+              error.message.eq(&Some(String::from(expected_err))),
+              "Got Err {:?}, but should have failed with message: {} for invalid_payloads.nth({})",
+              error.message,
+              expected_err,
+              idx
+            )
+          }
+        }
+      },
+    );
```

Comment:
I find it pretty confusing to loop through tests like this. Would be much clearer to have a helper function and call that explicitly with each set of params.

Reply:
@Nutomic Trying to understand your feedback here - are you saying you'd rather have a test per case, and call
```rust
match validate_update_payload(local_site_slur_filter_regex.clone(), false, true, edit_site)
{
  Ok(_) => {
    panic!(
    "Got Ok, but validation should have failed with error: {} for invalid_payloads.nth({})",
    expected_err, idx
  )
  }
  Err(error) => {
    assert!(
      error.message.eq(&Some(String::from(expected_err))),
      "Got Err {:?}, but should have failed with message: {} for invalid_payloads.nth({})",
      error.message,
      expected_err,
      idx
    )
  }
}
```
as the helper function?

Reply:
Now that you write it out, it doesnt seem to make much difference. Anyway the errors are being logged so it should be fine.

Reply:
I had started cleaning up the validation some more in anticipation of putting up a second PR, but at this point I just pushed it all up. I also updated the tests so that there's now a reason included in the tuples, which should make it easier to understand each case and not need to make a separate test for each one.